### PR TITLE
Add `ShardedTensor` support in `LightningModule`

### PR DESCRIPTION
**Pull request by yifuwang**
Register `ShardedTensor` state dict hooks in `LightningModule.__init__`
if the pytorch version supports `ShardedTensor`.

## What does this PR do?

Register `ShardedTensor` state dict hooks in `LightningModule.__init__` if the pytorch version supports `ShardedTensor` (#8633).

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]
FYI @SeanNaren @tchaton

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1974,3 +1976,13 @@ def __getstate__(self) -> Dict[str, Any]:
             state.pop("test_dataloader", None)
             state.pop("predict_dataloader", None)
         return state
+
+    def _register_sharded_tensor_state_dict_hooks_if_available(self) -> None:
+        """ """
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1981]: todo

File: tests/helpers/process_group.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+@contextmanager
+def single_process_pg():
+    """
+    Initialize the default process group with only the current process for
+    testing purposes. The process group is destoryed when the with block is
+    exited.
+    """
+    if dist.is_initialized():
+        raise RuntimeError(
+            "Can't use `single_process_pg_for_test` when the default process group is already initialized."
```
Comment (awaelchli) [inline tests/helpers/process_group.py:32]:
```suggestion
            "Can't use `single_process_pg ` when the default process group is already initialized."
```
was the name meant to match?

**Review by awaelchli — APPROVED** [review]

File: tests/helpers/process_group.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import torch.distributed as dist
+
+from pytorch_lightning.plugins.environments.lightning_environment import find_free_network_port
+
+
+@contextmanager
+def single_process_pg():
+    """
+    Initialize the default process group with only the current process for
+    testing purposes. The process group is destoryed when the with block is
```
Comment (SeanNaren) [inline tests/helpers/process_group.py:27]:
```suggestion
    testing purposes. The process group is destroyed when the with block is
```

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (SeanNaren) [inline pytorch_lightning/core/lightning.py:1981]:
```suggestion
        """
        Adds sharded tensor state dict hooks is sharded tensors are supported. These hooks ensure that 
        ShardedTensors are included when saving, and are loaded correctly. 
        """
```

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -1974,3 +1976,16 @@ def __getstate__(self) -> Dict[str, Any]:
             state.pop("test_dataloader", None)
             state.pop("predict_dataloader", None)
         return state
+
+    def _register_sharded_tensor_state_dict_hooks_if_available(self) -> None:
+        """
+        Adds sharded tensor state dict hooks is sharded tensors are supported. These hooks ensure that
+        ShardedTensors are included when saving, and are loaded correctly.
```
Comment (SeanNaren) [inline pytorch_lightning/core/lightning.py:1983]:
```suggestion
        Adds ShardedTensor state dict hooks if ShardedTensors are supported. These hooks ensure that
        ShardedTensors are included when saving, and are loaded the LightningModule correctly.
```

**Comment by justusschock** [comment]
@yifuwang @ananthsub do we need to do the same to the metrics modules in `torchmetrics`?

**Review by SeanNaren — APPROVED** [review]
Thanks for this, @yifuwang @ananthsub in the future would you see an API like this where we can handle distributed communications?

```python
class BoringModelWithShardedTensor(BoringModel):
    def __init__(self, spec):
        super().__init__()
        self.spec = spec

    def configure_sharded_model(self):
        self.sharded_tensor = dist._sharded_tensor.empty(self.spec, 10, 20)
        self.sharded_tensor.local_shards()[0].tensor.fill_(0)

model = BoringModelWithShardedTensor()
trainer = Trainer()
trainer.fit(model)
```

**Comment by yifuwang** [comment]
> @yifuwang @ananthsub do we need to do the same to the metrics modules in `torchmetrics`?

Good question. I'm not aware of any metric implementation that maintains sharded state. If there is, they should ideally be represented with sharded tensor once the feature's released, in which case this change also applies in `torchmetrics`.

**Comment by yifuwang** [comment]
> Thanks for this, @yifuwang @ananthsub in the future would you see an API like this where we can handle distributed communications?
> 
> ```python
> class BoringModelWithShardedTensor(BoringModel):
>     def __init__(self, spec):
>         super().__init__()
>         self.spec = spec
> 
>     def configure_sharded_model(self):
>         self.sharded_tensor = dist._sharded_tensor.empty(self.spec, 10, 20)
>         self.sharded_tensor.local_shards()[0].tensor.fill_(0)
> 
> model = BoringModelWithShardedTensor()
> trainer = Trainer()
> trainer.fit(model)
> ```

Very good question. I think it will be inevitable for `LightningModule` to recommend a best practice for organizing model parallel logic. However it's still not clear to me what it should look like.

I think the API above makes a lot of sense for model parallel uses cases that are developed on Lightning. However, for a self-contained model parallel use cases that just want to train with Lightning, there's a chance that the hook may be too opinionated.

It's definitely difficult to strike a good balance between usability and flexibility. I think it'd be helpful for us to do a informal survey of model parallel use cases we'd like to support in order to make an informed decision.

File: tests/core/test_lightning_module.py [hunk 5]
```diff
@@ -299,3 +303,39 @@ def assert_device(device: torch.device) -> None:
...
+
+class BoringModelWithShardedTensor(BoringModel):
+    def __init__(self, spec):
+        super().__init__()
+        self.sharded_tensor = dist._sharded_tensor.empty(spec, 10, 20)
+        self.sharded_tensor.local_shards()[0].tensor.fill_(0)
+
+
+@pytest.mark.skipif(
+    not _TORCH_SHARDED_TENSOR_AVAILABLE, reason="Test requires the torch version to support `ShardedTensor`"
```
Comment (Borda) [inline tests/core/test_lightning_module.py:316]: lest add argument for `RunIf`

Comment (awaelchli) [reply tests/core/test_lightning_module.py:316]: I think we should put `@RunIf(min_torch="1.10.0")` because this tensor type will appear in 1.10

File: tests/helpers/process_group.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+@contextmanager
+def single_process_pg():
+    """
+    Initialize the default process group with only the current process for
+    testing purposes. The process group is destroyed when the with block is
+    exited.
+    """
+    if dist.is_initialized():
+        raise RuntimeError("Can't use `single_process_pg ` when the default process group is already initialized.")
```
Comment (carmocca) [inline tests/helpers/process_group.py:31]:
```suggestion
        raise RuntimeError("Can't use `single_process_pg` when the default process group is already initialized.")
```

File: tests/helpers/process_group.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import os
+from contextlib import contextmanager
+
+import torch.distributed as dist
+
+from pytorch_lightning.plugins.environments.lightning_environment import find_free_network_port
+
+
+@contextmanager
+def single_process_pg():
```
Comment (carmocca) [inline tests/helpers/process_group.py:24]: This could be a `pytest.fixture. See: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/conftest.py

File: tests/core/test_lightning_module.py (hunk 5 above)
Comment (yifuwang) [reply tests/core/test_lightning_module.py:316]: @awaelchli I think if we put `@RunIf(min_torch="1.10.0")`, the test won't run for the nightly CI. Is that desirable?

Comment (awaelchli) [reply tests/core/test_lightning_module.py:316]: @yifuwang why so? I see in the CI we install  `1.10.0.dev20210816` and so if we put `min_torch=1.10.0` this condition should certainly hold up.

Comment (yifuwang) [reply tests/core/test_lightning_module.py:316]: @awaelchli I think dev releases come before the final release: https://www.python.org/dev/peps/pep-0440/#developmental-releases

```
>>> from packaging.version import Version
>>> Version("1.10.0.dev20210816") < Version("1.10.0")
True
```

File: tests/helpers/process_group.py (hunk 7 above)
Comment (yifuwang) [reply tests/helpers/process_group.py]: Very cool! Thanks for the suggestion!

File: tests/conftest.py [hunk 8]
```diff
@@ -106,3 +107,27 @@ class ThreadingHTTPServer(ThreadingMixIn, HTTPServer):
         server_thread.start()
         yield server.server_address
         server.shutdown()
+
+
+@pytest.fixture
+def single_process_pg():
+    """
+    Initialize the default process group with only the current process for
+    testing purposes. The process group is destoryed when the with block is
```
Comment (ananthsub) [inline tests/conftest.py:116]:
```suggestion
    testing purposes. The process group is destroyed when the with block is
```

**Comment by s-rog** [comment]
This is an edge case, but the version check passes for [1.10.0a0+ecc3718](https://github.com/pytorch/pytorch/commit/ecc3718) (the pytorch version in NGC 21.07) but the subsequent import fails.

This is probably the result of Nvidia's cherry picking, and I'm unsure if this is something we want to support?
